api/query/search.go

```diff
@@ -85,82 +85,112 @@ func (sh structuredSearchHandler) ServeHTTP(w http.ResponseWriter, r *http.Reque
+	runIDsStr := strings.Join(runIDStrs, ",")
+	r2.URL.RawQuery = q.Encode()
+
+	// Check if the query is a simple (empty/just True, or test name only) query
 	var isSimpleQ bool
 	{
 		if _, isTrueQ := rq.AbstractQuery.(True); isTrueQ {
 			isSimpleQ = true
 		} else if exists, isExists := rq.AbstractQuery.(AbstractExists); isExists && len(exists.Args) == 1 {
 			simpleQ, isSimpleQ = exists.Args[0].(TestNamePattern)
 		}
-		q := r.URL.Query()
 		for _, param := range []string{"interop", "subtests", "diff"} {
 			val, _ := shared.ParseBooleanParam(q, param)
 			isSimpleQ = isSimpleQ && (val == nil || !*val)
 		}
-	}
-
-	if !isSimpleQ {
-		ctx := sh.api.Context()
-		hostname := sh.api.GetServiceHostname("searchcache")
-		// TODO: This will not work when hostname is localhost (http scheme needed).
-		fwdURL, _ := url.Parse(fmt.Sprintf("https://%s/api/search/cache", hostname))
-		fwdURL.RawQuery = r.URL.RawQuery
-
-		logger := shared.GetLogger(ctx)
-		logger.Infof("Forwarding structured search request to %s: %s", hostname, string(data))
 
-		client := sh.api.GetHTTPClientWithTimeout(time.Second * 15)
-		req, err := http.NewRequest("POST", fwdURL.String(), bytes.NewBuffer(data))
+		// Check old summary files. If any can't be found,
+		// use the searchcache to aggregate the runs.
+		summariesValid, err := sh.validateSummaryVersions(r2.URL.Query(), logger)
 		if err != nil {
-			logger.Errorf("Failed to create request to POST %s: %v", fwdURL.String(), err)
-			http.Error(w, "Error connecting to search API cache", http.StatusInternalServerError)
-			return
+			logger.Debugf("Error checking summary file names: %v", err)
 		}
-		req.Header.Add("Content-Type", "application/json")
+		isSimpleQ = isSimpleQ && summariesValid
+	}
 
-		resp, err := client.Do(req)
+	// Use searchcache for a complex query or if old summary files exist.
+	if !isSimpleQ {
+		resp, err := sh.useSearchcache(w, r, data, logger)
 		if err != nil {
-			logger.Errorf("Error connecting to search API cache: %v", err)
 			http.Error(w, "Error connecting to search API cache", http.StatusInternalServerError)
-			return
-		}
-
-		if resp.StatusCode < 200 || resp.StatusCode >= 300 {
-			msg := fmt.Sprintf("Error from request: POST %s: STATUS %d", fwdURL.String(), resp.StatusCode)
-			errBody, err2 := ioutil.ReadAll(resp.Body)
-			if err2 == nil {
-				msg = fmt.Sprintf("%s: %s", msg, string(errBody))
-				resp.Body = ioutil.NopCloser(bytes.NewBuffer(errBody))
+		} else {
+			defer resp.Body.Close()
+			w.WriteHeader(resp.StatusCode)
+			_, err = io.Copy(w, resp.Body)
+			if err != nil {
+				logger.Errorf("Error forwarding response payload from search cache: %v", err)
 			}
-			logger.Errorf(msg)
-		}
-
-		defer resp.Body.Close()
-		w.WriteHeader(resp.StatusCode)
-		_, err = io.Copy(w, resp.Body)
-		if err != nil {
-			logger.Errorf("Error forwarding response payload from search cache: %v", err)
 		}
 		return
 	}
 
+	q = r.URL.Query()
+	q.Set("q", simpleQ.Pattern)
+	q.Set("run_ids", runIDsStr)
```

Comment:
I just realized above we use run_id for the query string parameter and here we use run_ids for the query string parameter. Is it possible to use only one? (Let me know if this might be a breaking change)

Reply:
This is something that I've had issues with for earlier changes. I didn't personally write this logic, and so it's confusing when `run_id` and `run_ids` are used, and to my knowledge they are not necessarily interchangeable. It might be worth filing an issue for investigation later.

Reply:
You're right! Apologies. Added https://github.com/web-platform-tests/wpt.fyi/issues/2957